### DISCUSSION TITLE
x-ref doesn't work with x-data in version 3

### DISCUSSION THREAD

**Discussion by NudistBeaaach** [alpinejs/alpine#4045] [7. Bugs (archived)] [answered]
Quick example, this works:
```html
<html>
<head>
    <script defer src="alpine/packages/alpinejs/dist/cdn.js"></script>
</head>
<body x-data="">
    <button @click="$refs.text.remove()">Remove Text</button>
    <span x-ref="text">Hello 👋</span>
</body>
</html>
```
But this doesn't:
```html
<html>
<head>
    <script defer src="alpine/packages/alpinejs/dist/cdn.js"></script>
</head>
<body x-data="">
    <button @click="$refs.text.remove()">Remove Text</button>
    <span x-data="" x-ref="text">Hello 👋</span>
</body>
</html>
```
with the `x-data` tag present it seems that the span is not included / detected in $refs variable.

**Comment by ekwoka** [comment, marked as answer]
refs are added to the nearest component root

This is the simplest way to prevent components from squashing eachother.

**Reply by bennadel** [reply]
This could be another place where _scoping_ could be value add. Imagine something like where both `x-data` and `x-ref` allow for a `:scope` value:

```html
<div x-data:parent>
	<div x-data x-ref:parent="childRef">
		<span x-text="( typeof parent.childRef )"></span>
	</div>
</div>
```

Just thinking out loud.

**Reply by SimoTod** [reply]
Not sure, X-ref usage is more then often a code smell. It only makes sense when dealing with external third party libraries or when developing complex reusable components.
With alpine, you shouldn't really access and manipulate html elements but you should manipulate the component data and let the html react.
In that way, you usually get simpler component and write less convoluted logic.
It's just my point of view, though.

**Reply by bennadel** [reply]
You could very well be right. In my case, I happened upon this while playing around with a Calendar component. And I needed a way to tell the calendar to jump to a given year (ex, `gotoYear(2025)`). To do this, I had something like this in my parent component (app):

```js
// An app-level method.
function gotoYear( year ) {
    // Reach down into the DOM to get a child element. Grab the "controller". Invoke that method.
    Alpine.$data( this.$el.querySelector( ".calendar" ) )
        .gotoYear( year )
    ;
}
```

This was just a code kata - I'm playing around trying to figure out the right patterns.

**Reply by ekwoka** [reply]
You can also, always use x-init to assign the element to a specific key in the component.

Or call a function with the el as an argument to do whatever with it.

I did also make a scope plugin. `@ekwoka/alpine-scope` to kind of play around.

**Reply by bennadel** [reply]
@ekwoka that's a good point. Coming from AngularJS, I always forget about the init method. I say that because AngularJS also had an `ng-init` directive; but, in their documentation, they explicitly talk about it as a toy function that shouldn't be used except in demos 😆 so, I have all that historical emotional baggage to cope with.

I've looked at your Scope plugin (I've actually looked at all your plugins). It's an interesting approach. You use a `Symbol` to store the scope `Map`. I've never actually used a Symbol before in programming. I assume you do this so that no one can accidentally overwrite it with another assignment. Or maybe it has something to do with the `Proxy` and iteratable properties?

**Reply by ekwoka** [reply]
Symbol are just guaranteed unique (unless you use Symbol.for but then why even use Symbol?)

So it's a way to make a MOSTLY private value. So anyone trying to tamper with it would need to be quite explicit in trying to do so, since you won't accidentally stumble upon symbol keys. They won't show up in most normal inspection of the object, and generally aren't enumerable. So they are pretty safe to slap all over other peoples objects without messing anything up, and not worrying about how things like proxies or other object references might interfere with some other more "private" data storing.

Makes them absolutely perfect for this use case. We can provide a magic to expose it in a controlled way as we might want to, since magics are handled a bit differently than the data stack itself.

Fun Fact: Symbol is the ONLY constructor in the JavaScript spec that CANNOT be called with `new`.

**Reply by bennadel** [reply]
Great explanation, thank you! I know that circular references can cause issues in Alpine.js. I know that we've had a previous conversation, maybe regarding the `$watch()` call, where it does a `JSON.stringify()` under the hood in order to setup a deep watch. I honestly don't remember the context of that conversation; but, I wonder if using a `Symbol` in something like that would have helped side-step the circular reference enumeration. Again ... just thinking out loud.

**Reply by ekwoka** [reply]
Specifically avoiding that issue, yes, JSON.stringify will not even touch a symbol key on the object at all.

the proxy will basically not expose it at all unless you have the symbol in hand and and try to directly access it.
